providers/base/tests/test_pipewire_test.py

```diff
@@ -0,0 +1,364 @@
+        self.assertEqual(None, self.vc._volume)
+
+    def test_bus_message_handler_name_error(self):
+        struct = Gst.Structure().new_from_string("test")
+        message = Gst.Message().new_element(None, struct)
+        message.type = Gst.MessageType.ELEMENT
+        bus = Gst.Bus()
+        self.gmh.bus_message_handler(bus, message)
+        self.assertEqual(0, self.analyzer.number_of_samples)
+        self.assertEqual(None, self.vc._volume)
+
+    def test_level_method_none(self):
+        self.gmh.level_method(1, self.pc, self.vc)
+        self.assertEqual(None, self.vc._volume)
+
+    @patch("pipewire_test.VolumeController._wpctl_output")
+    def test_level_method_zore(self, mock_wpctl):
+        self.vc.set_volume(0)
+        self.gmh.level_method(1, self.pc, self.vc)
+        self.assertEqual(0, self.vc._volume)
+
+    def test_spectrum_method(self):
+        self.gmh.spectrum_method(self.analyzer, [-0, -10])
+        self.assertEqual(0, self.analyzer.number_of_samples)
+
+    @patch("pipewire_test.parse_spectrum_message_structure")
+    def test_bus_message_handler_spectrum(self, mock_parse_spectrum_message):
+        self_mock = MagicMock()
+
+        message_mock = MagicMock()
+        message_mock.type = Gst.MessageType.ELEMENT
+        message_mock.get_structure().get_name.return_value = "spectrum"
+
+        mock_parse_spectrum_message.return_value = {
+            "magnitude": 1
+        }
+
+        GStreamerMessageHandler.bus_message_handler(self_mock,
+                                                    MagicMock(),
+                                                    message_mock)
+
+        self.assertTrue(self_mock.spectrum_method.called)
+
+    def test_bus_message_handler_level(self):
+        self_mock = MagicMock()
+
+        message_mock = MagicMock()
+        message_mock.type = Gst.MessageType.ELEMENT
+        message_mock.get_structure().get_name.return_value = "level"
+        message_mock.get_structure().get_value.return_value = [0]
+
+        GStreamerMessageHandler.bus_message_handler(self_mock,
+                                                    MagicMock(),
+                                                    message_mock)
+
+        self.assertTrue(self_mock.level_method.called)
+
+
+class GstAudioObjectTests(unittest.TestCase):
+
+    def test_init(self):
+        gao = GstAudioObject()
+        self.assertEqual("GstAudioObject", gao.class_name)
+
+
+class ParseSpectrumMessageStructureTests(unittest.TestCase):
+    ss = "spectrum, endtime=(guint64)5000000000,"\
+         " timestamp=(guint64)4900000000,"\
+         " stream-time=(guint64)4900000000,"\
+         " running-time=(guint64)4900000000,"\
+         " duration=(guint64)100000000,"\
+         " magnitude=(float){ -60, -60 };"
+    ans = '{"endtime": 5000000000, "timestamp": 4900000000,'\
+          ' "stream-time": 4900000000, "running-time": 4900000000,'\
+          ' "duration": 100000000, "magnitude": [-60, -60]}'
```

Comment:
Please use multiline strings

Reply:
@hanhsuan Hi! we have updated the [code review criteria](https://github.com/canonical/checkbox/blob/main/CONTRIBUTING.md#code-review-criteria--workflow)

Change my formatting related comments to suggestions as it's not the requested changes, please feel free to close it.

Reply:
This will break the hack parser, therefore it couldn't be modified like this.

Reply:
This is probably breaking the hack parser because it contains an extra tab. Probaby a good idea to follow this up with a more robust parser! (or just a `textwrap.dedent`)